idea/idea-fir/src/org/jetbrains/kotlin/idea/quickfix/fixes/AddExclExclCallFixFactories.kt

```diff
@@ -0,0 +1,121 @@
+                    // `::bar -> this!!::bar`
+                    psi to true
+                }
+            }
+
+            // `bar` -> `this!!.bar`
+            is KtNameReferenceExpression -> psi to true
+
+            // `bar()` -> `this!!.bar()`
+            is KtCallExpression -> psi to true
+
+            // `-foo` -> `-foo!!`
+            // NOTE: Unsafe unary operator call is reported as UNSAFE_CALL, _not_ UNSAFE_OPERATOR_CALL
+            is KtUnaryExpression -> psi.baseExpression to false
+
+            is KtBinaryExpression -> {
+                val receiver = if (KtPsiUtil.isInOrNotInOperation(psi)) {
+                    // `bar in foo` -> `bar in foo!!`
+                    psi.right
+                } else {
+                    // `foo + bar` -> `foo!! + bar` OR `foo infixFun bar` -> `foo!! infixFun bar`
+                    psi.left
+                }
+                receiver to false
+            }
+
+            // UNSAFE_INFIX_CALL/UNSAFE_OPERATOR_CALL on KtBinaryExpression is reported on the child KtOperationReferenceExpression
+            is KtOperationReferenceExpression -> return getFixForUnsafeCall(psi.parent)
+
+            else -> return emptyList()
+        }
+
+        return target.asFix(hasImplicitReceiver = hasImplicitReceiver)
+    }
+
+    // TODO: Move these factories to separate file when there are more *TYPE_MISMATCH fixes available.
+    // FE1.0 factory for TYPE_MISMATCH (QuickFixFactoryForTypeMismatchError) produces more than just AddExclExclFix.
+
+    val argumentTypeMismatchFactory = diagnosticFixFactory<KtFirDiagnostic.ArgumentTypeMismatch> { diagnostic ->
+        getFixesForTypeMismatch(diagnostic.psi, diagnostic.expectedType, diagnostic.actualType)
+    }
+
+    val returnTypeMismatchFactory = diagnosticFixFactory<KtFirDiagnostic.ReturnTypeMismatch> { diagnostic ->
+        getFixesForTypeMismatch(diagnostic.psi, diagnostic.expected, diagnostic.actual)
+    }
+
+    private fun KtAnalysisSession.getFixesForTypeMismatch(
+        psi: PsiElement,
+        expectedType: KtType,
+        actualType: KtType
+    ): List<IntentionAction> {
+        if (!expectedType.canBeNull && actualType.canBeNull) {
+            val nullableExpectedType = expectedType.makeNullable()
+            if (actualType isSubTypeOf nullableExpectedType) {
+                return psi.asFix()
+            }
+        }
+        return emptyList()
+    }
+
+    val iteratorOnNullableFactory = diagnosticFixFactory<KtFirDiagnostic.IteratorOnNullable> { diagnostic ->
+        val expression = diagnostic.psi as? KtExpression ?: return@diagnosticFixFactory emptyList()
+        val type = expression.getKtType()
+        if (!type.canBeNull) return@diagnosticFixFactory emptyList()
+
+        val hasValidIterator = type.getTypeScope()!!.getCallableSymbols { it == OperatorNameConventions.ITERATOR }
+            .filter { it is KtFunctionSymbol && it.isOperator && it.valueParameters.isEmpty() }.singleOrNull() != null
```

Comment:
I know FE1.0 does this. But it seems redundant for most of the cases. FIR checkers reporting IteratorOnNullable means there is already an iterator function. The only difference seems to be whether `operator` keyword is present. But TBH, I think the IDE should offer this fix even if the `operator` keyword is missing because this way, the IDE fix+diagnostics can natually lead user to discover the issue of missing `operator` keyword.

Reply:
I think that is fair. Right now I'm seeing the "Surround with null check" fix being offered in the same case even if it the `operator` keyword is missing. So I guess we can do the same here.

Reply:
Made this change. See 7th commit.